synapse/static/client/login/js/login.js

```diff
@@ -46,20 +49,37 @@ var setFeedbackString = function(text) {
 };
 
 var show_login = function() {
-    $("#loading").hide();
-
     var this_page = window.location.origin + window.location.pathname;
     $("#sso_redirect_url").val(this_page);
 
-    if (matrixLogin.serverAcceptsPassword) {
-        $("#password_flow").show();
+    if (matrixLogin.serverAcceptsCas) {
+        $("#sso_form").attr("action", "/_matrix/client/r0/login/cas/redirect");
     }
 
-    if (matrixLogin.serverAcceptsSso) {
-        $("#sso_flow").show();
-    } else if (matrixLogin.serverAcceptsCas) {
-        $("#sso_form").attr("action", "/_matrix/client/r0/login/cas/redirect");
-        $("#sso_flow").show();
+    if ((matrixLogin.serverAcceptsSso || matrixLogin.serverAcceptsCas)) {
+        if (try_token()) {
+            // Only show the SSO form if there's a login token in the query. That's
+            // because, if there is a token, and this function is run, it means an error
+            // happened, and in this case it's nicer to show the form with an error
+            // rather than redirect immediately to the SSO portal.
+            $("#sso_form").show();
+        } else {
+            // Submit the SSO form instead of displaying it. The reason behind this
+            // behaviour is that the user will likely arrive here after clicking on a
+            // button, in the client, with a label such as "Continue with SSO". And even
+            // if that's not the case, it kind of makes sense to direct the user directly
+            // to the SSO portal and skip the password login form.
```

Comment:
surely we shouldn't break password login for servers that support it? We should only skip straight to SSO if that's the only option supported by the server.

Reply:
Hmm, the goal of that was to match the UX of Riot clients, but I guess that's fair.

Reply:
Done in https://github.com/matrix-org/synapse/pull/7152/commits/45409c2ef3aa42a08ab18246038c735a40b5d394